airbyte-integrations/connectors/source-postgres/src/test/java/io/airbyte/integrations/source/postgres/XminPostgresSourceTest.java

```diff
@@ -0,0 +1,285 @@
+
+        ctx.fetch(
+            "CREATE TABLE names(first_name VARCHAR(200) NOT NULL, last_name VARCHAR(200) NOT NULL, power double precision NOT NULL, PRIMARY KEY (first_name, last_name));");
+        ctx.fetch(
+            "INSERT INTO names (first_name, last_name, power) VALUES ('san', 'goku', 'Infinity'),  ('prince', 'vegeta', 9000.1), ('piccolo', 'junior', '-Infinity');");
+        return null;
+      });
+    }
+  }
+
+  private static Database getDatabase(final DSLContext dslContext) {
+    return new Database(dslContext);
+  }
+
+  private static DSLContext getDslContext(final JsonNode config) {
+    return DSLContextFactory.create(
+        config.get(JdbcUtils.USERNAME_KEY).asText(),
+        config.get(JdbcUtils.PASSWORD_KEY).asText(),
+        DatabaseDriver.POSTGRESQL.getDriverClassName(),
+        String.format(DatabaseDriver.POSTGRESQL.getUrlFormatString(),
+            config.get(JdbcUtils.HOST_KEY).asText(),
+            config.get(JdbcUtils.PORT_KEY).asInt(),
+            config.get(JdbcUtils.DATABASE_KEY).asText()),
+        SQLDialect.POSTGRES);
+  }
+
+  private JsonNode getXminConfig(final PostgreSQLContainer<?> psqlDb, final String dbName) {
+    return Jsons.jsonNode(ImmutableMap.builder()
+        .put(JdbcUtils.HOST_KEY, psqlDb.getHost())
+        .put(JdbcUtils.PORT_KEY, psqlDb.getFirstMappedPort())
+        .put(JdbcUtils.DATABASE_KEY, dbName)
+        .put(JdbcUtils.SCHEMAS_KEY, List.of(SCHEMA_NAME))
+        .put(JdbcUtils.USERNAME_KEY, psqlDb.getUsername())
+        .put(JdbcUtils.PASSWORD_KEY, psqlDb.getPassword())
+        .put(JdbcUtils.SSL_KEY, false)
+        .put("replication_method", getReplicationMethod())
+        .build());
+  }
+
+  private JsonNode getReplicationMethod() {
+    return Jsons.jsonNode(ImmutableMap.builder()
+        .put("method", "Xmin")
+        .build());
+  }
+
+  @AfterAll
+  static void cleanUp() {
+    PSQL_DB.close();
+  }
+
+  @Test
+  void testDiscover() throws Exception {
+    final AirbyteCatalog actual = new PostgresSource().discover(getXminConfig(PSQL_DB, dbName));
+    actual.getStreams().forEach(actualStream -> {
+      final Optional<AirbyteStream> expectedStream =
+          CATALOG.getStreams().stream().filter(stream -> stream.getName().equals(actualStream.getName())).findAny();
+      assertTrue(expectedStream.isPresent());
+      assertEquals(expectedStream.get(), actualStream);
+    });
+  }
+
+  @Test
+  void testReadSuccess() throws Exception {
+    // Perform an initial sync with the configured catalog, which is set up to use xmin_replication.
+    // All of the records in the configured stream should be emitted.
+    final ConfiguredAirbyteCatalog configuredCatalog =
+        CONFIGURED_XMIN_CATALOG
+            .withStreams(CONFIGURED_XMIN_CATALOG.getStreams().stream().filter(s -> s.getStream().getName().equals(STREAM_NAME)).collect(
+                Collectors.toList()));
+    final List<AirbyteMessage> actualMessages =
```

Comment:
For my understanding, what is the initial state ?

Reply:
There are 3 reads (syncs) here : 
1. Sync 1 : no state (starting from scratch)
2. Sync 2 : State from sync 1 (the xmin that was calculated then)
3. Sync 3 : State from sync 2